service/history/workflow/mutable_state_impl.go

```diff
@@ -5113,26 +5113,18 @@ func (ms *MutableStateImpl) closeTransactionHandleActivityUserTimerTasks(
 	}
 }
 
-func (ms *MutableStateImpl) closeTransactionCollapseUpsertVisibilityTasks() {
-	// check if we have >= 2 tasks that are identical upsert visibility tasks
-	// note that VisibilityTimestamp and TaskID are not assigned yet
+func (ms *MutableStateImpl) closeTransactionCollapseVisibilityTasks() {
 	visTasks := ms.InsertTasks[tasks.CategoryVisibility]
 	if len(visTasks) < 2 {
 		return
 	}
-	var task0 *tasks.UpsertExecutionVisibilityTask
-	for i, task := range visTasks {
-		task, ok := task.(*tasks.UpsertExecutionVisibilityTask)
-		if !ok {
-			return
-		}
-		if i == 0 {
-			task0 = task
-		} else if *task != *task0 {
-			return
+	// Visibility tasks are collapsed into a single one: START < UPSERT < COLSE < DELETE
+	// Their enum values are already in order, so using them to make the code simpler.
+	for i := 1; i < len(visTasks); i++ {
+		if visTasks[i].GetType() >= visTasks[0].GetType() {
```

Comment:
should it check that the `Version` field matches? that was a comment on the PR where I added this... https://github.com/temporalio/temporal/pull/4727#discussion_r1284777021

I don't know what exactly it's used for myself

Reply:
I don't think version has any utility for visibility tasks. I saw some comments in the code that it's supposed to be the doc version for Elasticsearch, but the implementation actually uses the TaskID instead. cc: @alexshtin 

The version for the visibility task is the mutable state version at the moment the task was created, which, technically, should be in non-decreasing values as the order of the tasks added to the list of tasks here. But again, nothing is really preventing from tasks being added in a weird order (eg: the refresh was adding upsert after close with the same mutable state version).

Reply:
Maybe a better logic would be to keep the task with highest `Version`, and in case of ties collapse following the order START < UPSERT < CLOSE < DELETE?

Reply:
Actually, the version really doesn't really matter. There are some minor checks of the version (like the close task version matches the mutable state last write version). Besides that, the visibility task will be getting the latest data from mutable state to persist, so in this sense, keeping this collapse logic should work fine.

Reply:
> some minor checks of the version (like the close task version matches the mutable state last write version)

That's the main/original purpose of the version field I think, to make sure when the task is being processed, the mutable state loaded is still the mutable state which generates the workflow and there's no failover in between. 

- But now that I think about it, if we always get the visibility record from the current branch of the workflow and write to ES, that should also work and we don't really need the version check?  Maybe I missed something here. cc @alexshtin @yux0 if you can provide some more context.
- Within one mutable state transaction, the version is always the same? cc @yux0 ?

Reply:
> If we always get the visibility record from the current branch of the workflow and write to ES, that should also work and we don't really need the version check?

I think this is correct. Just want to bring it up to make sure I am not oversee anything. We also want to add the workflow state into SA (mainly because the zombie state). With this, I don't think it will change anything of the ^ statement. Because zombie workflows are from different runs. They are still the current branches.

> Within one mutable state transaction, the version is always the same?

Yes.

Reply:
Thanks @yux0 in that case I think we don't need the version check when collapsing.

cc @rodrigozhou re. adding workflow state into SA.